src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -216,17 +216,26 @@ let PostParseModuleSpec (_i, defaultNamespace, isLastCompiland, fileName, intf)
 
         SynModuleOrNamespaceSig(lid, isRecursive, kind, decls, xmlDoc, attributes, None, range, trivia)
 
-let GetScopedPragmasForHashDirective hd =
+let GetScopedPragmasForHashDirective hd (langVersion: LanguageVersion) =
+    let supportsNonStringArguments =
+        langVersion.SupportsFeature(LanguageFeature.ParsedHashDirectiveArgumentNonString)
+
     [
         match hd with
         | ParsedHashDirective("nowarn", numbers, m) ->
             for s in numbers do
-                match s with
-                | ParsedHashDirectiveArgument.SourceIdentifier _ -> ()
-                | ParsedHashDirectiveArgument.String(s, _, _) ->
-                    match GetWarningNumber(m, s) with
-                    | None -> ()
-                    | Some n -> ScopedPragma.WarningOff(m, n)
+                let warningNumber =
+                    match supportsNonStringArguments, s with
+                    | _, ParsedHashDirectiveArgument.SourceIdentifier _ -> None
+                    | true, ParsedHashDirectiveArgument.LongIdent _ -> None
+                    | true, ParsedHashDirectiveArgument.Int32(n, _) -> GetWarningNumber(m, string n, true)
+                    | true, ParsedHashDirectiveArgument.Ident(s, _) -> GetWarningNumber(m, s.idText, true)
+                    | _, ParsedHashDirectiveArgument.String(s, _, _) -> GetWarningNumber(m, s, true)
```

Comment:
Just thinking out loud, but:

This may be part of the reason that warnings like `IDE0059` are not honored by `#nowarn`, as `GetWarningNumber` always returns `None` for these.

Just thinking out loud: would it make sense to return something like a DU that maintains this info, so that we can actually process such other warnings? Thinking of something like:

```f#
type WarningNumber =
    | FSharpWarning of int
    | IdeWarning of int
    | NuGetWarning of int
    | Otherwarning of string // this allows us to keep the context, or use 'string * int' here?
```

Reply:
I don't know what IDE0059 is.  The F# compiler only supports FSn and n for nowarn.  We could have an issue to add FCS support for IDEn, but TBH, I don't understand why that wouldn't be FS0059, if the compiler generated it.

Reply:
We certainly wouldn't want to teach the compiler about warning prefixes or formats other than those it expects.  I.e. FSn and n.

Reply:
I agree with Kevin, that would invert the flow of dependencies.
IDE059 originates in the IDE layer as an IDE feature and core of the compiler should not be aware of it.

However, the IDE feature producing this could be enhanced to check file-level directives and filter what it produces accordingly.

Reply:
> 
> In C#, it is possible to switch these off. In F#, it is not.
> 

This gets triggered by a VS setting IIRC.
However, it is not impossible for the source of the diagnostic to respect "nowarn" pragmas.

Reply:
I think we can probably filter Non F# warning values in the build target if we wanted.  But it wouldn't really change anything in the compiler.  We would probably still not validate the strings any differently.

Reply:
> But... I am not even sure if this warning goes through the language processor... tbh, I've always been a little confused and frustrated about this one.

The warning from the build targets are supplied using --nowarn: blah on the command line. the validation for # nowarn: values and --nowarn values is the same code.  Which has the advantage of consistency and the disadvantage that ideally the #warn syntax would be more restrictive, because the compiler "knows" what it understands, whereas --nowarn: is at the mercy of sdk build targets.

Reply:
Thank you both for the insights and explanations. I guess the conclusion is: ignore my suggestion, it won't do what I originally thought it did. And perhaps we cab take a look at these IDE warnings some time. They are particularly confusing to newcomers. Others just ignore them.